onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -330,10 +330,29 @@ Status CUDAExecutionProvider::OnRunStart() {
   auto& current_deferred_release_event = GetPerThreadContext().GetCurrentDeferredReleaseEvent();
   CUDA_RETURN_IF_ERROR(cudaEventCreate(&current_deferred_release_event, cudaEventDisableTiming));
   deferred_release_cpu_ptr_.emplace(current_deferred_release_event, DeferredReleaseCPUPtrs());
+
+  if (ConfiguredForGraphCapture() && !IsGraphCaptured()) {
+    if (FinishWarmupForGraphCapture()) {
```

Comment:
Is it documented somewhere that we need to run multiple warmups **prior** to the graph capture for optimal performance ? 

Can the same be achieved by capturing the graph on the first Run() (like it was previously) and then the user having to run multiple GraphReplay Runs() instead as "warmups" ? That way we can avoid this ep configuration altogether.

Reply:
Synced offline. Warmup is needed, but `cuda_graph_warmup_runs` param is removed from `OrtCUDAProviderOptionsV2`. Instead we set a default value for it inside `CUDAExecutionProvider`.

Reply:
But can warmup just be one Run() prior to the capture? Here, the warm up run count is hard-coded to 3 and the user will have to know that she needs to perform 3 warm-ups prior to the graph replay (the way it is setup in the cuda graph python test code).

Can the logic be something like this:

'''
if (ConfiguredForGraphCapture() && !IsGraphCaptured()) {
    // Perform one regular Run() as warm-up
   
   CaptureBegin();
}
'''

Reply:
For the end user, they do not need to care about the concept of warmup or explicitly run warmups. They only need to enable cuda graph and then run session with io binding as below. The difference is that the first few runs may be slower because it did not use cuda graph. The multiple steps mentioned in the unit tests are to explain to developers about the internal steps at the backend, the end users do not need to run cuda graph in that way.
```
providers = [('CUDAExecutionProvider', {'enable_cuda_graph': True})]
session = onnxrt.InferenceSession(get_name("matmul_2.onnx"), providers=providers)
io_binding = session.io_binding()
...
session.run_with_iobinding(io_binding)
...
input.update_inplace(...)
session.run_with_iobinding(io_binding)
...
input.update_inplace(...)
session.run_with_iobinding(io_binding)
....
```
I agree that the logic you suggest is more straightforward than the current one. But there seems no clean way to perform one regular run inside `CUDAExecutionProvider`. On the other hand, in either way, the usages for end users will be almost the same.

Reply:
I see. But this arbitrary, first 'N' Runs() will not invoke cudaGraphs, but on the (N+1)th run the cudaGraph usage will be triggered will be a support nightmare. What will happen is people configure the CUDA EP to use cuda graphs and trigger an nvidia profile and notice that cuda graph usage was not triggered on the first 2 runs and open Github issues for this even if we document this.  

It may even be worth it if we can re-factor some code to make the suggestion happen rather than go with the current way because of the support tickets that may be filed because of this.

Reply:
Sure, I will change the warm-up count to be `1` and rename the variable at this moment. If other reviewers also think it will be an issue, I will make further change.

Reply:
That is fair. Thanks.

Reply:
Done